### PR TITLE
Define constants programmatically

### PR DESCRIPTION
**Pull request by fjarrett**
This is a partial implementation of https://github.com/wp-cli/ideas/issues/4. It doesn't include any tests (yet), or the ability to update variables - though I'm beginning to wonder if that should really be in the same command?

Anyhow, I'm not really sold on the `wp config define` semantics here. Mostly just want to see what others think of this direction and get some feedback. I needed this for myself and thought I'd post up here what I had come up with so far.

### CODE REVIEW

File: src/Config_Command.php [hunk 1]
```diff
@@ -163,6 +163,107 @@ public function create( $_, $assoc_args ) {
 	}
 
 	/**
+	 * Define a constant in the wp-config.php file.
+	 *
+	 * ## OPTIONS
+	 *
+	 * [<name>]
```
Comment (miya0001) [inline src/Config_Command.php]: `name` should be required?

File: src/Config_Command.php [hunk 2]
```diff
@@ -163,6 +163,107 @@ public function create( $_, $assoc_args ) {
...
+	 * : The constant name to be defined.
+	 *
+	 * [<value>]
```
Comment (miya0001) [inline src/Config_Command.php]: `value` should be required?

File: src/Config_Command.php [hunk 3]
```diff
@@ -163,6 +163,107 @@ public function create( $_, $assoc_args ) {
...
+	 * : The constant value to be defined.
+	 *
+	 * [--add]
```
Comment (miya0001) [inline src/Config_Command.php]: What should it return the status code when `--add` will be given and the `name` already exists?
This PR returns `0`, I am thinking it should return `1`.

File: src/Config_Command.php [hunk 4]
```diff
@@ -163,6 +163,107 @@ public function create( $_, $assoc_args ) {
...
+	 * : Add definition if it doesn't exist. Default: true
+	 *
+	 * [--update]
```
Comment (miya0001) [inline src/Config_Command.php]: `--update` has a similar issue with the `-add`.
I think we need more discussion about it.

**Comment by danielbachhuber** [comment]
Thanks for taking a pass on this @fjarrett.

From the usage perspective, I was thinking about two separate commands:

* `wp config set <type> <key> <value>` - Set a constant or a global to `wp-config.php`. `<type>` is either 'constant' or 'global', `<key>` is the name of the variable, and `<value>` is the value. If the constant or global doesn't yet exist, adds it to the `wp-config.php`
* `wp constant remove <type> <key>` - Remove a constant or global from `wp-config.php`.

From the implementation perspective, I think we'd be better off writing a standalone library (with lots of test coverage) for programmatically modifying a `wp-config.php` file. Once we have that in place, the specifics of the command implementation will be relatively straightforward (as it will be using the API provided by the library).

Would you like me to draft this library, or would you like the honors?

**Comment by fjarrett** [comment]
@danielbachhuber Yes please draft up what you're thinking and we can run from there.

**Comment by fjarrett** [comment]
@danielbachhuber I'm ready to work on this but am fuzzy on what the skeleton for a standalone library should look like. If you can draft sorta what you had in mind I can run with it.

**Comment by danielbachhuber** [comment]
@fjarrett I'm offline until Monday the 18th and planning to pick this back up after then. Want to join our office hours on Tuesday at 9 am Pacific?

**Comment by fjarrett** [comment]
@danielbachhuber Yes, see ya there! Have a good break man.

**Comment by danielbachhuber** [comment]
@fjarrett Here's a very basic first pass: https://github.com/wp-cli/wp-config-transformer/blob/master/src/WPConfigTransformer.php

I'll develop further when I have some time.

**Comment by danielbachhuber** [comment]
Closing in favor of working on the library first.

**Pull request closed without merging**